### PR TITLE
513 webpage measurement

### PR DESCRIPTION
**Pull request by jamesaj23**
Webpage download implemented in `honestybox-measurement`

A couple of different methods attempted for this one, using the `requests` and `pywebcopy` libraries. These implementations ended up running into `robots.txt` related issues, and would often download inconsistent numbers of files.

In the end shelling out to a variation of the existing `wget` implementation yielded the best results, with the `regex` on the output string thankfully not being particularly complicated. 

From the regex the following metrics are extracted
- `download_rate`
- `download_rate_unit` (forced by the `--report-speed=bits` flag to be in bits)
- `download_size`
- `download_size_unit`
- `downloaded_file_count`
- `elapsed_time`
- `elapsed_time_unit`

Following this, a `LatencyMeasurement` is also performed to the given host, with the results returned together as a list.

I decided configure the measurement to accept a single URL, rather than multiple and finding the least latent, as I don't really see the usefulness in possibly connecting to multiple different websites. Let me know if a usecase exists for that and it should be fairly easy to change.

### CODE REVIEW

**Comment by amorphic** [comment]
@jamesaj23 I'm a bit disappointed that we've ended up with another measurement that requires a command-line utility to be installed (`wget`). I'm really surprised that this wasn't possible using Python alone.

Can you elaborate on the problems that solutions such as `requests` + `beauitfulsoup` or `pywebcopy` had with `robots.txt`? In most cases I would have thought these could be avoided by spoofing the user agent (e.g. https://stackoverflow.com/a/36446917).

Given that this is complete and working I guess that we can use it as-is for now. But I'd like to understand how we might retain the interface but replace the underlying shell/`wget` mechanism sometime in future to remove the `wget` depednency.

File: measurement/plugins/ip_route/results_py3.py [hunk 1]
```diff
@@ -7,8 +7,7 @@
 
 @dataclass(frozen=True)
 class IPRouteMeasurementResult(MeasurementResult):
-    """Encapsulates the result from a IPRoute measurement.
-    """
+    """Encapsulates the result from a IPRoute measurement."""
```
Comment (amorphic) [inline measurement/plugins/ip_route/results_py3.py:10]: _an_ IPRoute measurement...

sorry

File: measurement/plugins/webpage_download/measurements.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
+import time
+import requests
+from six.moves.urllib.parse import urlparse
+from bs4 import BeautifulSoup
```
Comment (amorphic) [inline measurement/plugins/webpage_download/measurements.py:4]: I get an import error even after `poetry install`ing.

Don't add dependencies manually. Add them via `poetry add` to ensure they're added to `pyproject.toml` and the versions are locked in `poetry.lock`.

```
ImportError while importing test module '/home/james/dev/honestybox/honestybox-measurement/measurement/plugins/webpage_download/tests/test_measurements.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
measurement/plugins/webpage_download/tests/test_measurements.py:10: in <module>
    from measurement.plugins.webpage_download.measurements import WebpageMeasurement
measurement/plugins/webpage_download/measurements.py:4: in <module>
    from bs4 import BeautifulSoup
E   ModuleNotFoundError: No module named 'bs4'
```

File: measurement/plugins/webpage_download/measurements.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+            asset_count=len(to_download),
+            failed_asset_downloads=failed_asset_downloads,
+            elapsed_time=elapsed_time,
+            elapsed_time_unit=TimeUnit("s"),
+            errors=[],
+        )
+
+    def _parse_html(self, content):
+        soup = BeautifulSoup(content, "html.parser")
+        # hlinks = soup.find_all("a")
```
Comment (amorphic) [inline measurement/plugins/webpage_download/measurements.py:100]: nix the stray comment

File: measurement/plugins/webpage_download/measurements.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        failed_asset_downloads = 0
+        for asset in to_download:
+            try:
+                # Identify data URLs (already downloaded inline, counted in main download size)
+                if "data:" in asset:
+                    continue
+
+                # Check if path w/o preceeding slashes is a valid URL
+                if asset.startswith("//"):
+                    download_url = "https:" + asset
```
Comment (amorphic) [inline measurement/plugins/webpage_download/measurements.py:132]: Are we assuming that every asset will be ssl-secured here? What happens if an asset is not ssl-secured? Shouldn't we be using `http`|`https` based on the `asset` url or at least the `url` originally passed to the measurement?

**Review by amorphic — CHANGES_REQUESTED** [review]
Thanks for your patience while I get to this.

I'm unable to test yet as there's an `ImportError` as per comment.

Also a couple of nitpicks and a question regarding the `https` assumption. Once these are resolved I'll be able to test properly.

File: measurement/plugins/ip_route/results_py3.py (hunk 1 above)
Comment (jamesaj23) [reply measurement/plugins/ip_route/results_py3.py:10]: I think this one is actually correct (for once) :laughing: this change was due to a `black` update I believe

**Comment by amorphic** [comment]
FYI we only include `LatencyMeasurementResult`s in the results from other measurements because we _must_ perform a `LatencyMeasurement` as part of those measurements. As per your description we're measuring a specific URL here and so no such `LatencyMeasurement` is necessary. As such, I've removed it.

**Review by amorphic — APPROVED** [review]
I've tidied a few things up and added a catch for `requests.ConnectionError` (which is distinct from `ConnectionError` and was being raised when an invalid url was passed).

Good to go.

**Pull request merged**